src/System.Private.CoreLib/shared/System/Collections/Concurrent/ConcurrentQueue.cs

```diff
@@ -276,26 +276,29 @@ public int Count
         {
             get
             {
-                ConcurrentQueueSegment<T> head, tail;
-                int headHead, headTail, tailHead, tailTail;
+                // Take the cross-segment lock to prevent _head and _tail from changing while we're
+                // trying to compute the count.  Once we've taken the lock, we know that any internal
+                // segments (not the head and not the tail) won't have their counts changed, as you
+                // can't dequeue from or enqueue to an internal segment.  Thus, we only need to get
+                // a stable count for the head and tail segments, 
                 var spinner = new SpinWait();
                 while (true)
                 {
                     // Capture the head and tail, as well as the head's head and tail.
-                    head = _head;
-                    tail = _tail;
-                    headHead = Volatile.Read(ref head._headAndTail.Head);
-                    headTail = Volatile.Read(ref head._headAndTail.Tail);
+                    ConcurrentQueueSegment<T> head = _head;
+                    ConcurrentQueueSegment<T> tail = _tail;
+                    int headHead = Volatile.Read(ref head._headAndTail.Head);
+                    int headTail = Volatile.Read(ref head._headAndTail.Tail);
 
                     if (head == tail)
                     {
-                        // There was a single segment in the queue.  If the captured
-                        // values still (or again) represent reality, return the segment's
-                        // count. A single segment should be the most common case once the
-                        // queue's size has stabilized after segments have grown to
-                        // the point where growing is no longer needed.
+                        // There was a single segment in the queue.  If the captured segments still
+                        // match, then we can trust the values to compute the segment's count. (It's
+                        // theoretically possible the values could have looped around and still exactly match,
+                        // but that would required at least ~4 billion elements to have been enqueued and
+                        // dequeued between the reads.)
                         if (head == _head &&
-                            head == _tail &&
+                            tail == _tail &&
```

Comment:
Was the previous comparison here a mistake?

Reply:
No, head and tail contain the same value, as validated by the if statement above.  I just felt it was clearer and more obvious to use tail instead of head.